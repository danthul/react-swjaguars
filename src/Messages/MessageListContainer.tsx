import * as React from "react";
import * as moment from "moment";
import Panel from "../Panel";
import { getMessagesApi, deleteMessageApi } from "../apiHelpers";
import MessageList from "./MessageList";

type Props = {
  /* */
};

type messageItem = {
  body: string;
  name: string;
  updated: string;
  _id: string;
};

type State = {
  messages?: messageItem[];
  editMessage: messageItem;
  currentUser: any;
  editing: string;
};

class MessageListContainer extends React.Component<Props, State> {
  constructor(props: any) {
    super(props);
    this.state = {
      editMessage: {
        _id: "0",
        body: "",
        name: "",
        updated: moment(new Date()).format("lll")
      },
      currentUser: "",
      editing: ""
    };
  }

  newFN = () => {
    this.setState({ editing: "0" });
  };

  editFN = (id: string) => {
    this.setState({ editing: id });
    //go find this message and set editMessage
    if (this.state.messages) {
      const editMessage = this.state.messages.find(message => {
        return message._id === id;
      });
      if (editMessage) {
        this.setState({ editMessage });
      }
    }
    return true;
  };

  deleteFN = (id: string) => {
    if (confirm("Are you sure you wish to delete this message?")) {
      deleteMessageApi(id)
        .then(() => {
          if (this.state.messages) {
            const otherMessages = this.state.messages.filter(message => {
              return message._id !== id;
            });
            if (otherMessages) {
              this.setState({ messages: otherMessages });
            }
          }
        })
        .catch(response => {
          console.error(response);
          alert("Something went wrong");
        });
    }
    return true;
  };

  editComplete = (editedMessage: messageItem) => {
    //find message and replace with this
    if (this.state.messages && editedMessage) {
      const messageID = this.state.messages.findIndex(message => {
        return message._id === editedMessage._id;
      });
      if (messageID > -1) {
        // this.setState;
        const cloneMessages = [...this.state.messages];
        cloneMessages[messageID] = editedMessage;
        this.setState({
          messages: cloneMessages
        });
      } else {
        this.setState({
          messages: [editedMessage, ...this.state.messages]
        });
      }
    }
    this.setState({ editing: "" });
  };

  async componentDidMount() {
    try {
      const response = await getMessagesApi();
      if (response && response.data) {
        this.setState({ messages: response.data });
      }
    } catch (e) {
      console.error(e); // ðŸ’©
    }
  }

  render() {
    return (
      <MessageList
        editing={this.state.editing}
        editMessage={this.state.editMessage}
        editComplete={this.editComplete}
        newFN={this.newFN}
        editFN={this.editFN}
        deleteFN={this.deleteFN}
        messages={this.state.messages}
      />
    );
  }
}

export default MessageListContainer;
