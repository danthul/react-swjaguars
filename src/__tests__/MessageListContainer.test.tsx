import * as React from "react";
// import * as ReactDOM from "react-dom";
import * as axios from "axios";
import * as MockAdapter from "axios-mock-adapter";
// import { spy } from "sinon";
import { configure, mount } from "enzyme";
import toJson from "enzyme-to-json";
import MessageListContainer from "../Messages/MessageListContainer";
// setup file
import * as Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

// This sets the mock adapter on the default instance
let mock = new MockAdapter(axios);
let instance;

beforeEach(function() {
  instance = axios;
  mock = new MockAdapter(instance);
});

const messages = [
  {
    _id: "5a22c1d5fa2bdd4d546428bc",
    body: "another new one",
    name: "New message #2",
    __v: 0,
    updated: "2017-12-02T15:07:00.000Z"
  },
  {
    _id: "5a1f58999675c43714acd671",
    body: "Yo yo\n",
    name: "New message",
    __v: 0,
    updated: "2017-11-30T01:02:00.000Z"
  }
];

mock.onGet("/articles").reply(200, messages);

mock.onGet("/articles/last").reply(200, messages[0]);

describe("MessageListContainer", () => {
  it("gets successful response from API", async () => {
    const item = mount(<MessageListContainer />);
    return instance.get("/articles").then(response => {
      expect(toJson(item.update())).toMatchSnapshot();
      const divCount = item.find("div").length;
      expect(item.find("div").length).toBeGreaterThan(0);
      const bodyDiv = item.find(".panel-body");
      expect(document.querySelectorAll(".panel-body").length).toEqual(3);
    });
  });
});
