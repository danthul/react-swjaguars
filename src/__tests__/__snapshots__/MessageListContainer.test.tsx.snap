// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageListContainer gets successful response from API 1`] = `
<MessageListContainer>
  <MessageList
    deleteFN={[Function]}
    editComplete={[Function]}
    editFN={[Function]}
    editMessage={
      Object {
        "_id": "0",
        "body": "",
        "name": "",
        "updated": "2017-12-11T10:09:23.794-06:00",
      }
    }
    editing=""
    newFN={[Function]}
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <main
          className="col-md-12"
        >
          <div />
          <Panel>
            <div
              className="panel panel-default"
            >
              <div
                className="panel-body"
              >
                <button
                  className="btn btn-primary"
                  onClick={[Function]}
                >
                  Click here to create a new message
                </button>
              </div>
            </div>
          </Panel>
        </main>
      </div>
    </div>
  </MessageList>
</MessageListContainer>
`;
